File: practice/algorithms/implementation/strange_counter/StrangeCounter.scala
object StrangeCounter extends App {
  val sc = new java.util.Scanner (System.in)
  val t = sc.nextLong()

  val cycle = Math.sqrt((t.toDouble / 3).ceil).ceil.toLong
  val cycleEndTime = cycleEnd(t, 0, 3)
  val timeDiff = cycleEndTime - t
  val finalValue = timeDiff + 1

  println(finalValue)

  def cycleEnd(time: Long, start: Long, step: Long): Long = {
    if(start > time) {
      return start
    }
    cycleEnd(time, start + step, step * 2)
  }
}

File: practice/algorithms/implementation/strange_counter/StrangeCounter.java
import java.util.*;

public class StrangeCounter {

    public static void main(String[] args) {
        Scanner stdin = new Scanner(System.in);
        long t = stdin.nextLong();
        long cycleEndTime = cycleEnd(t, 0, 3);
        long timeDiff = cycleEndTime - t;
        long finalValue = timeDiff + 1;

        System.out.println(finalValue);
        stdin.close();
    }

    private static long cycleEnd(long time, long start, long step) {
        if(start > time) {
            return start;
        }
        return cycleEnd(time, start + step, step * 2);
    }
}